### PR TITLE
Climb

### PR DESCRIPTION
**Pull request by OliverW10**

### CODE REVIEW

File: components/hanger.py [hunk 1]
```diff
@@ -1,3 +1,48 @@
...
+from utilities.ctre import FALCON_CPR
+from magicbot import tunable
+
+
 class Hanger:
+
+    climb_motor: ctre.TalonFX
+
+    GEAR_RATIO = 1 / 35.0
+    PULLEY_CIRCUMFERANCE = 0.04
```
Comment (auscompgeek) [inline components/hanger.py:11]: typo:

```suggestion
    PULLEY_CIRCUMFERENCE = 0.04
```

File: components/hanger.py [hunk 2]
```diff
@@ -1,3 +1,48 @@
...
+        self.climb_motor.config_kF(0, 0.0, 10)
+        self.climb_motor.config_kP(0, 2.0, 10)
+        self.climb_motor.config_kI(0, 0.0, 10)
+        self.climb_motor.config_kD(0, 0.0, 10)
+        self.climb_motor.configSelectedFeedbackSensor(
+            ctre.FeedbackDevice.IntegratedSensor, 0, 10
+        )
+        self.climb_motor.setNeutralMode(ctre.NeutralMode.Brake)
+        self.climb_motor.configSelectedFeedbackCoefficient(
+            self.GEAR_RATIO * self.PULLEY_CIRCUMFERANCE / FALCON_CPR, 0, 10
```
Comment (auscompgeek) [inline components/hanger.py:29]:
```suggestion
            self.GEAR_RATIO * self.PULLEY_CIRCUMFERENCE / FALCON_CPR, 0, 10
```

File: components/hanger.py [hunk 3]
```diff
@@ -1,3 +1,48 @@
...
+    climb_motor: ctre.TalonFX
+
+    GEAR_RATIO = 1 / 35.0
+    PULLEY_CIRCUMFERANCE = 0.04
+
+    winch_dir = 0
+    winch_power = tunable(0)
+
+    def setup(self) -> None:
+        # TODO: check
```
Comment (auscompgeek) [inline components/hanger.py:17]: Is this done?
```suggestion
```

File: robot.py [hunk 4]
```diff
@@ -164,6 +168,16 @@ def teleopPeriodic(self) -> None:
                 self.indexer_control.wants_to_intake = True
                 self.intake.deployed = True
 
+        # TODO: change buttons to codriver controller
```
Comment (auscompgeek) [inline robot.py:171]: What needs to happen before this can be done?

**Review by auscompgeek — APPROVED** [review]

File: robot.py [hunk 5]
```diff
@@ -31,6 +32,7 @@ class MyRobot(magicbot.MagicRobot):
     led_control: LedController
     shooter_control: ShooterController
     indexer_control: IndexerController
+    climb_control: ClimbController
```
Comment (auscompgeek) [inline robot.py:35]: Doesn't this need to run before the `shooter_controller`?

File: components/intake.py [hunk 6]
```diff
@@ -7,11 +7,12 @@ class Intake:
     intake_motor: rev.CANSparkMax
     intake_piston: wpilib.DoubleSolenoid
 
-    deployed = False
     auto_retract = True
 
-    def __init__(self) -> None:
+    def __init__(self):
```
Comment (auscompgeek) [inline components/intake.py:12]:
```suggestion
    def __init__(self) -> None:
```

File: controllers/climb.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+class ClimbController(StateMachine):
+    """High level hang controller"""
+
+    shooter_control: ShooterController
+    turret: Turret
+    intake: Intake
+    hanger: Hanger
+
+    @state(first=True, must_finish=True)
+    def prepare(self, state_tm):
```
Comment (auscompgeek) [inline controllers/climb.py:18]:
```suggestion
    def prepare(self, state_tm: float) -> None:
```

File: controllers/climb.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        self.shooter_control.track_target = False
+        self.turret.slew_local(math.pi)
+        self.intake.auto_retract = False
+        self.intake.motor_enabled = False
+        self.intake.deployed = True
+        if state_tm > 1 and self.turret.is_on_target():
+            self.next_state("raise_arms")
+
+    @timed_state(duration=5, must_finish=True)
+    def raise_arms(self):
```
Comment (auscompgeek) [inline controllers/climb.py:28]:
```suggestion
    def raise_arms(self) -> None:
```

**Review by auscompgeek — COMMENTED** [review]
Apparently I started another review and forgot to submit, whoops.

These suggestions are non-blocking, except for the robot component ordering.

File: components/hanger.py [hunk 9]
```diff
@@ -1,3 +1,48 @@
...
+        )
+        self.winch_dir = 0
+
+    def winch(self) -> None:
+        self.winch_dir = -1
+
+    def payout(self) -> None:
+        self.winch_dir = 1
+
+    def is_raised(self) -> None:
```
Comment (auscompgeek) [inline components/hanger.py:47]: Fixing the CI:
```suggestion
    def is_raised(self) -> bool:
```

File: robot.py [hunk 10]
```diff
@@ -176,6 +171,18 @@ def teleopPeriodic(self) -> None:
         if self.codriver.getAButton():
             self.chassis.set_pose_failsafe()
 
+        # Climb
+        if self.codriver.getYButton():
```
Comment (james-ward) [inline robot.py:175]: In the past we have used two simultaneous button presses to activate things like this - accidental activation during the game would be a disaster!
Can we get back to regular play if this controller is activated? Is it worth putting in that ability?

**Review by LucienMorey — APPROVED** [review]

**Pull request merged**